### PR TITLE
chore: bump vault_tls library to 0.12 (tls-certificates v4)

### PR DESCRIPTION
**Pull request by gruyaume**
# Description

Bump the `vault_tls` library to 0.12 in order to use the tls certificates library v4.

Superseeds #275 

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/charms/vault_k8s/v0/vault_tls.py [hunk 1]
```diff
@@ -136,91 +150,146 @@ def __init__(
...
         super().__init__(charm, "tls")
         self.charm = charm
         self.workload = workload
         self._service_name = service_name
         self.tls_directory_path = tls_directory_path
-        self.tls_access = TLSCertificatesRequiresV3(charm, TLS_CERTIFICATE_ACCESS_RELATION_NAME)
+        self.common_name = common_name
+        self.sans_dns = sans_dns
+        self.sans_ip = sans_ip
+        self.mode = self._get_mode()
```
Comment (DanielArndt) [inline lib/charms/vault_k8s/v0/vault_tls.py:165]: [minor] It feels a little odd to set this at the class level when it's only used immediately after.

I wonder if using a property, or even a local variable makes more sense.

This implementation is a lot easier to follow, though, so I like that.

File: lib/charms/vault_k8s/v0/vault_tls.py [hunk 2]
```diff
@@ -499,42 +471,47 @@ def generate_vault_ca_certificate() -> Tuple[str, str]:
...
+        ca=Certificate.from_string(ca_certificate),
+        ca_private_key=PrivateKey.from_string(ca_private_key),
         csr=csr,
         validity=365 * 50,
     )
-    return vault_certificate.decode()
+    return str(vault_private_key), str(vault_certificate)
+
+
+def existing_certificate_is_self_signed(ca_certificate: Certificate) -> bool:
```
Comment (DanielArndt) [inline lib/charms/vault_k8s/v0/vault_tls.py:515]:
```suggestion
def certificate_is_self_signed(ca_certificate: Certificate) -> bool:
```

**Review by DanielArndt — CHANGES_REQUESTED** [review]

**Review by DanielArndt — CHANGES_REQUESTED** [review]
The name change (`existing_certificate_is_self_signed`) seems worthwhile, my other comment is more an observation and I wouldn't mind if nothing changed.

**Comment by gruyaume** [comment]
> The name change (`existing_certificate_is_self_signed`) seems worthwhile, my other comment is more an observation and I wouldn't mind if nothing changed.

The comments are relate to a PR in the vault k8s lib which was merged a couple of days ago. Is the recommendation to make the changes there and port them here before merging this?

**Comment by DanielArndt** [comment]
> > The name change (`existing_certificate_is_self_signed`) seems worthwhile, my other comment is more an observation and I wouldn't mind if nothing changed.
> 
> @DanielArndt 
> The comments in your review relate to a PR in the vault k8s lib which was merged a couple of days ago. Here we simply pull this lib and use it as it is. Is the recommendation to make the changes there and port them here before merging this?
> 
> Reference:
> - https://github.com/canonical/vault-k8s-operator/pull/463

Oops, no. I thought these were the changes to make it compatible with the new lib. My bad.

**Review by DanielArndt — APPROVED** [review]

**Pull request merged**